libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/IDPManager.kt

```diff
@@ -207,50 +224,99 @@ internal class IDPManager(
+
+            activeFlow?.authCodeActivity?.let {authCodeActivity ->
+                getAuthCode(context, authCodeActivity, currentUser, spConfig, message)
+            } ?: run {
+                val intent = message.toIntent().apply {
+                    putExtra(SRC_APP_PACKAGE_NAME_KEY, spConfig.appPackageName)
+                    // Intent action needs to be ACTION_VIEW, so passing message action through extras
+                    putExtra(ACTION_KEY, message.action)
+                    setAction(Intent.ACTION_VIEW)
+                    setClass(context, com.salesforce.androidsdk.auth.idp.IDPAuthCodeActivity::class.java)
+                    addCategory(Intent.CATEGORY_DEFAULT)
+                }
+                // The activity will call onReceive which will call handleLoginRequest but this time
+                // with a IDPAuthCodeActivity attached to the flow
+                startActivity(context, intent)
+            }
+        } ?: run {
+            send(context, IDPToSPResponse(message.uuid, error = "IDP app not logged in"), spConfig.appPackageName)
+        }
+    }
+
+    /**
+     * Get auth code for the SP app
+     */
+    fun getAuthCode(
+        context: Context,
+        authCodeActivity: IDPAuthCodeActivity,
+        currentUser: UserAccount,
+        spConfig: SPConfig,
+        message: SPToIDPRequest
+    ) {
+        activeFlow?.onStatusUpdate?.let { it(Status.GETTING_AUTH_CODE_FROM_SERVER) }
+        sdkMgr.generateAuthCode(
+            authCodeActivity,
+            currentUser,
+            spConfig,
+            message.codeChallenge
+        ) { result ->
+            result.error?.let {error ->
+                // We failed to get an auth code
+                if (activeFlow?.firstMessage is IDPToSPRequest) {
+                    // We are in a IDP initiated flow - let the IDP app know
+                    activeFlow?.let { it.onStatusUpdate(Status.ERROR_RECEIVED_FROM_SERVER) }
+                } else {
+                    // We are NOT in a IDP initiated flow - we need to let the SP app know
+                    val response = IDPToSPResponse(message.uuid, error = error)
+                    send(context, response, spConfig.appPackageName)
                 }
-            )
+            } ?: run {
+                // We successfully got an auth code - send it to the SP app
+                val response = IDPToSPResponse(message.uuid, code = result.code, loginUrl = result.loginUrl)
+                send(context, response, spConfig.appPackageName)
+                // Let the IDP app know
+                activeFlow?.let { it.onStatusUpdate(Status.AUTH_CODE_SENT_TO_SP) }
+            }
+        }
+    }
 
+    /**
+     * Attach authCodeActivity to IDP manager's active flow (setting up the flow first if needed)
+     */
+    fun attachToActiveFlow(
+        context: Context,
+        authCodeActivity: IDPAuthCodeActivity,
+        spAppPackageName: String?
+    ) {
+
+        activeFlow?.let { flow ->
```

Comment:
We need an active flow and the auth code activity "attached" to it for code generation to be able to get its hands on the web view (and also finish the activity when code generation is completed).